mgmt/utils/MgmtMarshall.h

```diff
@@ -61,17 +75,112 @@ struct MgmtMarshallData {
-ssize_t mgmt_message_write_v(int fd, const MgmtMarshallType *fields, unsigned count, va_list ap);
-
-ssize_t mgmt_message_parse(const void *ptr, size_t len, const MgmtMarshallType *fields, unsigned count, ...);
-ssize_t mgmt_message_parse_v(const void *ptr, size_t len, const MgmtMarshallType *fields, unsigned count, va_list ap);
-
-ssize_t mgmt_message_marshall(void *ptr, size_t len, const MgmtMarshallType *fields, unsigned count, ...);
-ssize_t mgmt_message_marshall_v(void *ptr, size_t len, const MgmtMarshallType *fields, unsigned count, va_list ap);
+/// A byte to idenify the marshalled type.
+static constexpr uint8_t MGMT_INT_TYPE    = 0x00;
+static constexpr uint8_t MGMT_LONG_TYPE   = 0x01;
+static constexpr uint8_t MGMT_STRING_TYPE = 0x02;
+static constexpr uint8_t MGMT_DATA_TYPE   = 0x03;
+
+/**
+  Various marshalling functions are implemented below. All functions are written
+  recursively with variadic templates and overloads for specific MgmtMarshall types.
+
+  The benefit of this is that the marshall functions are all type-safe at compile time
+  because an invalid type will not resolve to a overload function. However, because
+  there is no generic template definition, we need to overload each function with an
+  empty version to allow for correct expansion.
+  */
+
+/// mgmt_message_length -------------------------------------------------------
+MgmtMarshallInt mgmt_message_length(MgmtMarshallInt *field);
+MgmtMarshallInt mgmt_message_length(MgmtMarshallLong *field);
+MgmtMarshallInt mgmt_message_length(MgmtMarshallString *field);
+MgmtMarshallInt mgmt_message_length(MgmtMarshallData *field);
+MgmtMarshallInt mgmt_message_length(); // to allow for expansion
+
+template <typename T, typename... Rest>
+MgmtMarshallInt
+mgmt_message_length(T first, Rest &&... rest)
+{
+  MgmtMarshallInt nfirst = mgmt_message_length(first);
+  MgmtMarshallInt nrest  = mgmt_message_length(std::forward<Rest>(rest)...);
+  return (nfirst != -1 && nrest != -1) ? nfirst + nrest : -1;
+}
+/// end mgmt_message_length ---------------------------------------------------
+
+/// mgmt_message_read ---------------------------------------------------------
+ssize_t mgmt_message_read(int fd, MgmtMarshallInt *field);
+ssize_t mgmt_message_read(int fd, MgmtMarshallLong *field);
+ssize_t mgmt_message_read(int fd, MgmtMarshallString *field);
+ssize_t mgmt_message_read(int fd, MgmtMarshallData *field);
+ssize_t mgmt_message_read(int fd); // to allow for expansion
+
+template <typename T, typename... Rest>
+ssize_t
+mgmt_message_read(int fd, T first, Rest &&... rest)
+{
+  ssize_t nfirst = mgmt_message_read(fd, first);
+  ssize_t nrest  = mgmt_message_read(fd, std::forward<Rest>(rest)...);
+  return (nfirst != -1 && nrest != -1) ? nfirst + nrest : -1;
+}
+/// mgmt_message_read ---------------------------------------------------------
+
+/// mgmt_message_write --------------------------------------------------------
```

Comment:
It's unclear to me why there's both `..._write` and `..._marshal`. If the message is marshalled to a buffer, what is the use of per type write methods?

Reply:
So you have the option of either marshaling into a buffer or directly writing on the socket. It's more for flexibility. 

For example, `NetworkMessage.h` marshalls the whole message and sends it as a data object. The point of this is to attach a length so the entire marshalled message can be read into a buffer. This is necessary because the entire buffer is passed to the corresponding handler which figures out how to parse the buffer. Contrast this to a response which can directly send the parameters using `mgmt_message_write`.